File: python/self/other/valid-parentheses.py
class Solution:
    def valid(self, s: str) -> bool:
        stack = []
        map = {"]": "[", "}": "{", ")": "("}
        for char in s:
            if char in map.values():
                # 遇到左括号 -> 直接入栈
                stack.append(char)
            elif char in map.keys():
                # 遇到右括号 -> 如果栈空或栈顶不是对于的左括号 -> 返回False
                if stack == [] or map[char] != stack.pop():
                    return False
            else:
                # 不是题目所定义的符号
                return False
        return stack == []

File: cpp/category/other/valid-parentheses.cpp
#include <stack>
#include <string>

using namespace std;
class Solution
{
public:
    bool isValid(string s)
    {
        stack<char> parent;
        for (char &c : s)
        {
            switch (c)
            {
            case '(':
            case '{':
            case '[':
                parent.push(c);
                break;
            case ')':
                if (parent.empty() || parent.top() != '(')
                    return false;
                else
                    parent.pop();
                break;
            case '}':
                if (parent.empty() || parent.top() != '{')
                    return false;
                else
                    parent.pop();
                break;
            case ']':
                if (parent.empty() || parent.top() != '[')
                    return false;
                else
                    parent.pop();
                break;
            default:; // pass
            }
        }
        return parent.empty();
    }

private:
    bool isValid(string s)
    {
        stack<char> stk;
        for (const auto &c : s)
        {
            switch (c)
            {
            case '{':
                stk.push('}');
                break;
            case '[':
                stk.push(']');
                break;
            case '(':
                stk.push(')');
                break;
            default:
                if (stk.size() == 0 || c != stk.top())
                    return false;
                else
                    stk.pop();
            }
        }
        return stk.size() == 0;
    }
};